### PR TITLE
neigh: Support multi device neighbor discovery

### LINKED ISSUE #19908: Packets to LB dropped due to FIB lookup failed
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We see an issue that packets to LB sometimes are dropped on intermediate nodes because of FIB lookup failed. Our loadBalancer setup is loadBalancer.mode=dsr and algorithm=maglev. 

```
$ kubectl -n kube-system exec cilium-hpw7g -- cilium monitor -t drop
xx drop (FIB lookup failed) flow 0xe19b1328 to endpoint 0, , identity unknown->unknown: 10.0.0.19:42178 -> 10.64.5.180:8443 tcp SYN
```

There is no L2 neighbor entry for its next hop in the arp table. I think it causes the FIB lookup failed on intermediate nodes.

```
$ ip r get 10.64.5.180 from 10.0.0.19 iif eno1
10.64.5.180 from 10.0.0.19 via 10.69.0.133 dev eno2
    cache iif eno1
$ ip r get 10.64.5.180 from 10.0.0.19 iif eno2
10.64.5.180 from 10.0.0.19 via 10.69.0.133 dev eno2
    cache iif eno2

$ ip n show 10.69.0.133
// None
```

Our nodes have two devices connected to a different L3 network, and we attach a global scope address to a dummy interface. The bird running on nodes advertises routes to the global scope address to peer ToRs. 

```
$ ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
2: eno1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc mq state UP group default qlen 1000
    link/ether <MAC_ADDRESS> brd ff:ff:ff:ff:ff:ff
    inet 10.69.0.68/26 brd 10.69.0.127 scope link eno1
       valid_lft forever preferred_lft forever
3: eno2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1460 qdisc mq state UP group default qlen 1000
    link/ether <MAC_ADDRESS> brd ff:ff:ff:ff:ff:ff
    inet 10.69.0.132/26 brd 10.69.0.191 scope link eno2
       valid_lft forever preferred_lft forever
4: node0: <BROADCAST,NOARP,UP,LOWER_UP> mtu 1500 qdisc noqueue state UNKNOWN group default qlen 1000
    link/ether <MAC_ADDRESS> brd ff:ff:ff:ff:ff:ff
    inet 10.69.0.4/32 scope global node0
       valid_lft forever preferred_lft forever
```

A route entry for another node is like the following(from node 10.69.0.4 to 10.69.0.5). Its next-hop is either 10.69.0.69 or 10.69.0.133. Cilium's L2 neighbor discovery seems to cover only one next-hop. I suspect that the current L2 neighbor discovery doesn't support our node network setup like this. 

```
$ ip r
10.69.0.5 proto bird metric 32
        nexthop via 10.69.0.69 dev eno1 weight 1
        nexthop via 10.69.0.133 dev eno2 weight 1

$ ip n show 10.69.0.69
10.69.0.69 dev eno1 lladdr ${MAC_ADDRESS} extern_learn  REACHABLE
$ ip n show 10.69.0.133
// None
```

### Cilium Version

v1.11.5

### Kernel Version

```
$ uname -a
Linux rack0-cs4 5.15.37-flatcar #1 SMP Wed May 4 13:53:25 -00 2022 x86_64 AMD EPYC 7413 24-Core Processor AuthenticAMD GNU/Linux
```

### Kubernetes Version

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"clean", BuildDate:"2021-12-16T08:38:33Z", GoVersion:"go1.16.12", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.5", GitCommit:"5c99e2ac2ff9a3c549d9ca665e7bc05a3e18f07e", GitTreeState:"archive", BuildDate:"2022-03-23T08:04:34Z", GoVersion:"go1.17.8", Compiler:"gc", Platform:"linux/amd64"}
```

### Sysdump

My sysdump is too big to upload

### Relevant log output

_No response_

### Anything else?

Here is our values.yaml.

```
bgp:
  announce:
    loadbalancerIP: true
  enabled: true
bpf:
  hostRouting: true
cni:
  chainingMode: "generic-veth"
  customConf: true
datapathMode: "veth"
devices: "eth+ eno1+ eno2+"
enableIPv4Masquerade: false
enableIdentityMark: false
externalIPs:
  enabled: true
hostPort:
  enabled: false
hostServices:
  enabled: true
  hostNamespaceOnly: true
hubble:
  relay:
    enabled: true
    tls:
      server:
        enabled: true
    rollOutPods: true
    resources:
      requests:
        cpu: 100m
        memory: 200Mi
  tls:
    auto:
      method: "cronJob"
  metrics:
    enabled:
      - "dns"
      - "drop:destinationContext=pod|dns|ip;sourceContext=pod|dns|ip"
      - "tcp"
      - "flow:destinationContext=pod|dns|ip;sourceContext=pod|dns|ip"
      - "icmp"
      - "http"
k8sServiceHost: 127.0.0.1
k8sServicePort: 16443
kubeProxyReplacement: "partial"
loadBalancer:
  # We can't enable XDP Acceleration because of the issue https://github.com/cilium/cilium/issues/19453
  acceleration: disabled
  algorithm: maglev
  mode: dsr
maglev:
  hashSeed: 3HCx6JennjWtot2U
nodePort:
  directRoutingDevice: "e+"
  enabled: true
operator:
  rollOutPods: true
  prometheus:
    enabled: true
  resources:
    requests:
      cpu: 100m
      memory: 200Mi
policyAuditMode: false
policyEnforcementMode: "default"
prometheus:
  enabled: true
resources:
  requests:
    cpu: 100m
    memory: 400Mi
rollOutCiliumPods: true
tunnel: "disabled"
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by ysksuzuki**
Please see the commit message.

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

Fixes: #19908

### CODE REVIEW

**Comment by brb** [comment]
@ysksuzuki Thanks for the PR! Pinging @borkmann to review too, as he recently completely reworked the neigh management.

**Review by borkmann — APPROVED** [review]
Great work, thanks for addressing this limitation!

Could you please also extend the documentation under ...
 - https://docs.cilium.io/en/latest/gettingstarted/kubeproxy-free/#neighbor-discovery
 ... and document the updated behavior for multi-device in detail (maybe new paragraph for it)? Are there potential limitations, I presume it currently does not work yet with the runtime device detection, right?

**Comment by ysksuzuki** [comment]
@borkmann Thank you for the review!

> Could you please also extend the documentation under

Sure, will do that.

> I presume it currently does not work yet with the runtime device detection, right?

Right, it doesn't work for now. I think neighDiscoveryLink should follow the device changes here. Will do it in a follow-up PR later.
https://github.com/cilium/cilium/blob/aa7572be0f7cfece0ebfb739716ab970c478d358/daemon/cmd/daemon.go#L1287

**Comment by ysksuzuki** [comment]
@borkmann Updated the document for multi-device. Please check it.

File: Documentation/gettingstarted/kubeproxy-free.rst
Comment (qmonnet) [inline Documentation/gettingstarted/kubeproxy-free.rst]:
```suggestion
and global scope addresses each (10.69.0.1/26 and 10.69.0.2/26). A next-hop from node1 to node2 is
```

File: Documentation/gettingstarted/kubeproxy-free.rst
Comment (qmonnet) [inline Documentation/gettingstarted/kubeproxy-free.rst]:
```suggestion
::
```

File: Documentation/gettingstarted/kubeproxy-free.rst
Comment (qmonnet) [inline Documentation/gettingstarted/kubeproxy-free.rst]:
```suggestion
With, on node1:
    
    .. code-block:: shell-session
```

**Review by qmonnet — CHANGES_REQUESTED** [review]
A few suggestions on doc formatting, but this looks good, thank you!

**Comment by ysksuzuki** [comment]
@qmonnet Thank you for the review! I modified the document accordingly.

**Review by qmonnet — APPROVED** [review]
Thanks a lot! If you could only squash the last two commits together (docs fixes with docs update), it would be perfect :)

**Comment by ysksuzuki** [comment]
> if you could only squash the last two commits together

Done!

**Comment by ysksuzuki** [comment]
@borkmann @qmonnet Hello, is there anything I could do for this PR?

**Comment by ysksuzuki** [comment]
Two tests failed. Hitting known flakes?

ConformanceAKS: https://github.com/cilium/cilium/issues/19472
Multicluster: https://github.com/cilium/cilium/issues/20186 ?

**Comment by sayboras** [comment]
ConformanceAKS is disabled now due to high probability of failure. Multicluster should be fine now.

**Comment by brb** [comment]
@sayboras Is it good to merge?

**Comment by sayboras** [comment]
> Is it good to merge?

Yes it is, just wait for one more codeowner review for the bot to mark ready to merge.

**Review by NikAleksandrov — APPROVED** [review]

**Pull request merged**
